### ISSUE TITLE
Contour should set a status message when an ingressroute references an invalid secret

### ISSUE THREAD

**Issue by pickledrick** [heptio/contour#1169]
**What steps did you take and what happened:**

- created a secret of type `kubernetes.io/dockerconfigjson`
- referenced unexpected secret type in `Ingress` specification for TLS configuration
- received error message in logs

```
contour-6bdd96f88-ftrfp envoy [2019-06-13 05:03:08.569][000001][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_mux_subscription_lib/common/config/grpc_mux_subscription_impl.h:70] gRPC config for type.googleapis.com/envoy.api.v2.Listener rejected: Proto constraint validation failed (ListenerValidationError.FilterChains[i]: ["embedded message failed validation"] | caused by FilterChainValidationError.TlsContext: ["embedded message failed validation"] | caused by DownstreamTlsContextValidationError.CommonTlsContext: ["embedded message failed validation"] | caused by CommonTlsContextValidationError.TlsCertificates[i]: ["embedded message failed validation"] | caused by TlsCertificateValidationError.CertificateChain: ["embedded message failed validation"] | caused by field: "specifier", reason: is required): name: "ingress_https"
```

**What did you expect to happen:**
Error message in logs
```
secret namespace/name was malformed and unable to be loaded as a TLS listener
```

**Environment:**

- Contour version: `v0.11.0`
- Kubernetes version: (use `kubectl version`): `v1.14.2`
- Kubernetes installer & version: `kind v0.4.0-alpha`
- Cloud provider or hardware configuration: `local`
- OS (e.g. from `/etc/os-release`): `osx`

<hr/>

Blocked:
- [ ] All the issues we have about status

**Comment by davecheney** [comment]
Thanks for raising this issue. I'll write a regression test for this now.

Can you please confirm for me:

1. Which version of Envoy is the customer using?
2. Does 0.12.1 fix the problem?

**Comment by davecheney** [comment]
@pickledrick thanks for raising this issue. I write up some regression tests in #1170 but was not able to reproduce the issue you reported. As far as I can tell, even going back to 0.11, Contour has always checked that the `tls.crt` and `tls.key` keys are present in the secret before Contour will consider it valid.

Are you able to share some more details about the issue that was reported to you so I can help get to the bottom of how that invalid secret made it through to Envoy.

**Comment by pickledrick** [comment]
Specific steps to reproduce
ingress-secret.yaml
```
cat <<EOF > ingress.yaml
---
apiVersion: v1
data:
  .dockerconfigjson: ewogICAgImF1dGhzIjogewogICAgICAgICJodHRwczovL2luZGV4LmRvY2tlci5pby92MS8iOiB7CiAgICAgICAgICAgICJhdXRoIjogImMzUi4uLnpFMiIKICAgICAgICB9CiAgICB9Cn0K
kind: Secret
metadata:
  name: regcred
type: kubernetes.io/dockerconfigjson
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  labels:
    app: kuard
  name: kuard
spec:
  backend:
    serviceName: kuard
    servicePort: 80
  rules:
  - host: some-app.company.net
  tls:
  - hosts:
    - some-app.company.net
    secretName: regcred
EOF
```

- `kind create cluster`
- `export KUBECONFIG="$(kind get kubeconfig-path --name="kind")"`
- ` git clone git@github.com:heptio/contour.git`
- `cd contour`
- `git checkout release-0.11`
- `kubectl apply -f deployment/deployment-grpc-v2/`
- `kubectl apply -f https://j.hept.io/contour-kuard-example`
- `kubectl -n heptio-contour scale deploy contour --replicas=1`
- `kubectl apply -f ingress.yaml`
- `kubectl -n heptio-contour logs -f <pod-name> envoy | grep warning`

The warning error is returned from envoy, however contour does not supply a meaningful error.

**Comment by pickledrick** [comment]
Customer is using  envoy `v1.9.1`, testing contour `v0.12.1`

**Comment by pickledrick** [comment]
Using the `release-0.12` release branch to deploy, envoy still at `v1.9.1` doesn't return the same error? Not too sure why. Perhaps different configuration? And no error is returned from contour.

**Comment by davecheney** [comment]
Thanks for the update. 0.11 did no validation of the secret, as long as there was something that matched the name, contour would feed it to envoy. This was fixed in 0.12 and has been strengthened to be even stricter in 0.13 (due next week)

Given than versions of Contour < 0.12.1 should not be used, can we call this issue closed?

**Comment by shabx** [comment]
@davecheney , I brought your question with @pickledrick - Peter confirmed that checking in Contour versions > 0.12.1 is more strict and Contour and does not pass the object to Envoy so we don't see the error returned from Envoy.

We have updated the ticket and told the customer if this is good enough. It looks like they were looking for a proper error message in the logs when there is a malformed secret reference. We can't close the issue yet.

**Comment by davecheney** [comment]
Thanks for the update. 

The problem with reporting an error on a malformed secret is where to do so,

Logging to contour’s stdout is the obvious place but this has several drawbacks. 

- those logs may not be accessible to the end user, they may not have permission to tail the logs of a container in the heptio-contour namespace. 
- these logs are likely to be verbose as a naïve implementation will emit a long line every time contour goes through a configuration rebuild cycle — read potentially any time something related to ingress changes in the api server. 

The better place to put this information would be in the status field of a k8s object. The problem is secret has no status field that I know of. This leaves the ingress or ingressroute object itself. I think this is the best place to report a malformed secret. 

Sadly status updating has been a sore point for contour for a long time. We plan to fix it, but this is blocked on work scheduled for 0.14 so the soonest we could deliver this feature is 0.15.

**Comment by shabx** [comment]
Thanks @davecheney . I'll update the customer.

**Comment by davecheney** [comment]
I believe this was fixed in 0.14 by #1251 

```
  virtualhost:
    fqdn: httpbin2.davecheney.com
    tls:
      secretName: missing
status:
  currentStatus: invalid
  description: TLS Secret [missing] not found or is malformed
```

**Comment by systemlogic** [comment]
@davecheney @pickledrick How you have updated the cluster to get rid of the problem as mentioned in https://github.com/projectcontour/contour/issues/1169#issuecomment-526055946.
Can I have specific command/steps without impacting much clients cluster

**Comment by davecheney** [comment]
@systemlogic this issue was resolved in 0.14. Can you please upgrade to 1.0, the current stable release and open a new issue if the problem persists. Thanks

**Comment by systemlogic** [comment]
@davecheney Can I have steps/Command to do so ?

**Comment by davecheney** [comment]
We have an upgrade guide for recent versions of contour

https://projectcontour.io/resources/upgrading/

If you’re using an older version please consult the releases page 

https://github.com/projectcontour/contour/releases

**Comment by systemlogic** [comment]
I have give up. 
I have updated images as mentioned in my cluster. I have also setup new cluster and installed contour using 
`kubectl create -f https://raw.githubusercontent.com/projectcontour/contour/release-1.0/examples/render/contour.yaml`
created secret using self sign crt and pem
and then create ingressRoute resource
```
apiVersion: contour.heptio.com/v1beta1
kind: IngressRoute
metadata:
  creationTimestamp: "2019-11-16T12:28:04Z"
  generation: 5
  name: development
  namespace: default
  resourceVersion: "2594"
  selfLink: /apis/contour.heptio.com/v1beta1/namespaces/default/ingressroutes/development
  uid: ff81ee23-ff29-4c8d-be70-be62ad702827
spec:
  tcpproxy:
    services:
    - name: development
      port: 80
  virtualhost:
    fqdn: development.monorepo.com
    tls:
      secretName: monorepo
status:
  currentStatus: invalid
  description: TLS Secret [monorepo] not found or is malformed
```
and service 
```
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: 2019-11-15T15:41:40Z
  labels:
    jobId: 77de3d1d9ed62ef289e45efc6484ea35
    name: development
    serviceKind: microservice
    toolchain: cedp-jsonnet
  name: development
  namespace: default
  resourceVersion: "3901295"
  selfLink: /api/v1/namespaces/default/services/development
  uid: 46b2b412-7a62-41e2-98e0-f02a2c02488c
spec:
  clusterIP: 172.21.176.208
  ports:
  - name: http
    port: 443
    protocol: TCP
    targetPort: 80
  selector:
    jobId: 77de3d1d9ed62ef289e45efc6484ea35
  sessionAffinity: None
  type: ClusterIP
```
NOTE : I AM ABLE TO SEE APPLICATION DEFAULT PAGE WITH HTTP, HOWEVER USING tls with secretName is always giving me said error. PLEASE HELP.

Is there a proper documentation or an issue is required to fix it.

**Comment by davecheney** [comment]
Is the monorepos secret present in the default namespace? Does it have the correct “kubernetes.io/type”? Is it valid? Sadly the secret object does not have a status field we can report failures on so we must report that data to the contour processes standard error.
